File: sdks/typescript/src/v2/local.ts
// SPDX-License-Identifier: Apache-2.0

import { ClientError, ProductError, productError, type RequestOptions, type Response, type Transport } from "./models.js";
import {
  FRAME_HEADER_SIZE,
  FRAME_KIND,
  blake3,
  decodeEnd,
  decodeFrame,
  decodeProductError,
  decodeProductResponse,
  decodeWelcome,
  encodeAuthenticatedHello,
  encodeCancel,
  encodeFrame,
  encodeHello,
  encodeProductRequest,
  encodeWindowUpdate,
} from "./protocol.js";

/** Runtime-injected exact byte stream for AF_UNIX sockets or Windows named pipes. */
export interface LocalByteStream {
  readExact(length: number, signal?: AbortSignal): Promise<Uint8Array>;
  write(encoded: Uint8Array, signal?: AbortSignal): Promise<void>;
  close(): Promise<void> | void;
}

export type LocalConnector = (endpoint: string, signal?: AbortSignal) => Promise<LocalByteStream>;

export interface LocalTransportOptions {
  readonly clientIdentity?: string;
  readonly apiKey?: string;
  readonly maximumPending?: number;
}

/**
 * Native-local transport entry point.
 *
 * Node runtimes supply a connector backed by `node:net` for AF_UNIX paths and
 * bare Windows named-pipe namespaces. The connector carries exact HYPHLCL1 bytes;
 * it must not add length prefixes, JSON, or another wrapper protocol.
 */
export class LocalTransport implements Transport {
  readonly #endpoint: string;
  readonly #connector: LocalConnector;
  readonly #clientIdentity: string;
  #apiKey: Uint8Array | undefined;
  #stream: LocalByteStream | undefined;
  #nextStreamId = 1;
  #active: Promise<void> = Promise.resolve();
  #initialWindow = 64 * 1024;
  #maximumFramePayload = 16 * 1024 * 1024;
  #negotiatedMinor: number | undefined;
  readonly #maximumPending: number;
  #pending = 0;
  #closed = false;

  constructor(endpoint: string, connector: LocalConnector, options: LocalTransportOptions | string = {}) {
    const { clientIdentity = "hyphae-typescript-sdk-v2", apiKey, maximumPending = 64 } =
      typeof options === "string" ? { clientIdentity: options } : options;
    if (endpoint.length === 0) throw new ClientError("local endpoint must not be empty");
    if (typeof connector !== "function") throw new ClientError("a platform local connector is required");
    if (clientIdentity.length === 0 || new TextEncoder().encode(clientIdentity).byteLength > 4096) throw new ClientError("local client identity is invalid");
    if (!Number.isSafeInteger(maximumPending) || maximumPending <= 0 || maximumPending > 4096) throw new ClientError("local pending request bound is invalid");
    this.#endpoint = endpoint;
    this.#connector = connector;
    this.#clientIdentity = clientIdentity;
    if (apiKey !== undefined) {
      const encoded = new TextEncoder().encode(apiKey);
      encodeAuthenticatedHello(encoded, clientIdentity);
      this.#apiKey = encoded;
    }
    this.#maximumPending = maximumPending;
  }

  async execute(
    operation: string,
    args: Readonly<Record<string, unknown>>,
    options: RequestOptions = {},
  ): Promise<Response> {
    if (this.#closed) throw new ClientError("local transport is closed");
    if (this.#pending >= this.#maximumPending) throw new ClientError("local pending request queue is full");
    this.#pending += 1;
    const prior = this.#active;
    let release = (): void => {};
    this.#active = new Promise<void>((resolve) => { release = resolve; });
    let admitted = false;
    try {
      await waitForTurn(prior, options);
      admitted = true;
      return await this.#executeSerial(operation, args, options);
    } finally {
      this.#pending -= 1;
      if (admitted) release();
      else void prior.then(release, release);
    }
  }

  async close(): Promise<void> {
    if (this.#closed) return;
    this.#closed = true;
    const stream = this.#stream;
    this.#stream = undefined;
    this.#negotiatedMinor = undefined;
    this.#apiKey?.fill(0);
    this.#apiKey = undefined;
    await stream?.close();
  }

  async #disconnect(): Promise<void> {
    const stream = this.#stream;
    this.#stream = undefined;
    this.#negotiatedMinor = undefined;
    await stream?.close();
  }

  async #executeSerial(
    operation: string,
    args: Readonly<Record<string, unknown>>,
    options: RequestOptions,
  ): Promise<Response> {
    const requestId = checkedRequestId(options.requestId);
    if (options.signal?.aborted === true) throw productError("cancelled", requestId);
    if (options.deadlineMicros !== undefined && options.deadlineMicros <= BigInt(Date.now()) * 1000n) {
      throw productError("deadline_exceeded", requestId);
    }
    const deadlineController = new AbortController();
    const cancelDeadline = (): void => deadlineController.abort(options.signal?.reason);
    options.signal?.addEventListener("abort", cancelDeadline, { once: true });
    const deadlineTimer = localDeadlineTimer(options.deadlineMicros, deadlineController);
    let stream: LocalByteStream | undefined;
    let streamId: number | undefined;
    try {
      const handshakeId = requestId === (1n << 64n) - 1n ? 1n : requestId + 1n;
      const terminalReplay = operation === "security_api_key_revoke_self"
        || operation === "security_api_key_rotate_self_activate";
      const freshTerminalReplay = terminalReplay && this.#stream === undefined;
      const terminalPayload = freshTerminalReplay
        ? encodeProductRequest(operation, args, options, 3)
        : undefined;
      stream = await this.#connection(
        handshakeId,
        deadlineController.signal,
        terminalPayload === undefined
          ? undefined
          : { streamId: 1, requestId, payload: terminalPayload },
      );
      if (terminalPayload !== undefined) this.#nextStreamId = 2;
      streamId = this.#nextStreamId;
      this.#nextStreamId = this.#nextStreamId === 0xffffffff ? 1 : this.#nextStreamId + 1;
      const payload = encodeProductRequest(operation, args, options, this.#negotiatedMinor);
      const kind = operation === "sql_prepare" ? FRAME_KIND.prepare : operation === "sql_deallocate" ? FRAME_KIND.deallocate : FRAME_KIND.execute;
      if (!freshTerminalReplay) {
        await this.#writeFrame(stream, kind, streamId, requestId, payload, deadlineController.signal);
      }
      const chunks: Uint8Array[] = [];
      let length = 0;
      let credited = 0;
      const maximum = options.limits?.maxResponseBytes ?? 16 * 1024 * 1024;
      for (;;) {
        if (deadlineController.signal.aborted) {
          throw deadlineController.signal.reason ?? new ClientError("local request interrupted");
        }
        const frame = await readFrame(stream, this.#maximumFramePayload, deadlineController.signal);
        if (frame.streamId !== streamId || frame.requestId !== requestId) {
          await this.#disconnect();
          throw new ClientError("local response correlation mismatch");
        }
        if (frame.kind === FRAME_KIND.failure) throw new ProductError(decodeProductError(frame.payload));
        if (frame.kind === FRAME_KIND.data) {
          length += frame.payload.byteLength;
          if (length > maximum) throw new ClientError("local response exceeds the configured maximum");
          chunks.push(frame.payload);
          credited += frame.payload.byteLength;
          if (credited >= Math.max(1, Math.floor(this.#initialWindow / 2))) {
            await this.#writeFrame(stream, FRAME_KIND.windowUpdate, streamId, requestId, encodeWindowUpdate(BigInt(credited)), deadlineController.signal);
            credited = 0;
          }
          continue;
        }
        if (frame.kind === FRAME_KIND.end) {
          const encoded = join(chunks, length);
          const completion = decodeEnd(frame.payload);
          if (completion.totalBytes !== BigInt(length) || !equal(completion.digest, blake3(encoded))) {
            throw new ClientError("local provisional response completion mismatch");
          }
          return decodeProductResponse(encoded, requestId, this.#negotiatedMinor);
        }
        await this.#disconnect();
        throw new ClientError("local server returned an invalid response frame");
      }
    } catch (cause) {
      if (deadlineController.signal.aborted) {
        if (stream !== undefined && streamId !== undefined) {
          await this.#writeFrame(stream, FRAME_KIND.cancel, streamId, requestId, encodeCancel()).catch(() => {});
        }
        await this.#disconnect().catch(() => {});
        throw productError(Boolean(options.signal?.aborted) ? "cancelled" : "deadline_exceeded", requestId);
      }
      if (stream !== undefined) {
        await this.#disconnect().catch(() => {});
      }
      throw cause;
    } finally {
      if (deadlineTimer !== undefined) clearTimeout(deadlineTimer);
      options.signal?.removeEventListener("abort", cancelDeadline);
    }
  }

  async #connection(
    requestId: bigint,
    signal?: AbortSignal,
    terminalRequest?: { readonly streamId: number; readonly requestId: bigint; readonly payload: Uint8Array },
  ): Promise<LocalByteStream> {
    if (this.#stream !== undefined) return this.#stream;
    const connecting = this.#connector(this.#endpoint, signal);
    let stream: LocalByteStream;
    try {
      stream = await abortable(connecting, signal);
    } catch (error) {
      if (signal?.aborted === true) {
        void connecting.then((late) => late.close()).catch(() => {});
      }
      throw error;
    }
    try {
      const hello = this.#apiKey === undefined
        ? encodeHello(this.#clientIdentity, 3)
        : encodeAuthenticatedHello(this.#apiKey, this.#clientIdentity, 3);
      try {
        await abortable(stream.write(encodeFrame(FRAME_KIND.hello, 0, requestId, hello), signal), signal);
        if (terminalRequest !== undefined) {
          await abortable(
            stream.write(
              encodeFrame(
                FRAME_KIND.execute,
                terminalRequest.streamId,
                terminalRequest.requestId,
                terminalRequest.payload,
              ),
              signal,
            ),
            signal,
          );
        }
      } finally {
        hello.fill(0);
      }
      const frame = await readFrame(stream, 16 * 1024 * 1024, signal);
      if (frame.kind === FRAME_KIND.failure) throw new ProductError(decodeProductError(frame.payload));
      if (frame.kind !== FRAME_KIND.welcome || frame.streamId !== 0 || frame.requestId !== requestId) {
        throw new ClientError("local handshake response mismatch");
      }
      const welcome = decodeWelcome(frame.payload);
      if (this.#apiKey !== undefined && (BigInt(welcome.capabilities ?? 0) & 0x80n) === 0n) {
        throw new ClientError("local server downgraded managed API-key authentication");
      }
      this.#negotiatedMinor = Number(welcome.minor);
      this.#initialWindow = Number(welcome.initialWindow);
      const maximumFramePayload = Number(welcome.maximumFramePayload);
      if (!Number.isInteger(maximumFramePayload) || maximumFramePayload <= 0 || maximumFramePayload > 16 * 1024 * 1024) {
        throw new ClientError("local handshake frame limit is invalid");
      }
      this.#maximumFramePayload = maximumFramePayload;
      this.#stream = stream;
      return stream;
    } catch (error) {
      await stream.close();
      throw error;
    }
  }

  async #writeFrame(
    stream: LocalByteStream,
    kind: number,
    streamId: number,
    requestId: bigint,
    payload: Uint8Array,
    signal?: AbortSignal,
  ): Promise<void> {
    if (payload.byteLength > this.#maximumFramePayload) {
      throw new ClientError("local frame exceeds the negotiated maximum");
    }
    await abortable(stream.write(encodeFrame(kind, streamId, requestId, payload), signal), signal);
  }

}

async function waitForTurn(prior: Promise<void>, options: RequestOptions): Promise<void> {
  if (options.signal?.aborted === true) throw productError("cancelled", checkedRequestId(options.requestId));
  if (options.deadlineMicros !== undefined && options.deadlineMicros <= BigInt(Date.now()) * 1000n) {
    throw productError("deadline_exceeded", checkedRequestId(options.requestId));
  }
  const controller = new AbortController();
  const cancel = (): void => controller.abort("cancelled");
  options.signal?.addEventListener("abort", cancel, { once: true });
  const timer = localDeadlineTimer(options.deadlineMicros, controller);
  try {
    await Promise.race([
      prior,
      new Promise<never>((_, reject) => controller.signal.addEventListener("abort", () => {
        reject(productError(options.signal?.aborted === true ? "cancelled" : "deadline_exceeded", checkedRequestId(options.requestId)));
      }, { once: true })),
    ]);
  } finally {
    if (timer !== undefined) clearTimeout(timer);
    options.signal?.removeEventListener("abort", cancel);
  }
}

function localDeadlineTimer(
  deadlineMicros: bigint | undefined,
  controller: AbortController,
): ReturnType<typeof setTimeout> | undefined {
  if (deadlineMicros === undefined) return undefined;
  const remaining = deadlineMicros - BigInt(Date.now()) * 1000n;
  return setTimeout(() => controller.abort(), Math.max(0, Number((remaining + 999n) / 1000n)));
}

async function readFrame(stream: LocalByteStream, maximumPayload: number, signal?: AbortSignal) {
  const header = await abortable(stream.readExact(FRAME_HEADER_SIZE, signal), signal);
  const length = new DataView(header.buffer, header.byteOffset).getUint32(24, true);
  if (length > maximumPayload) throw new ClientError("local frame exceeds the negotiated maximum");
  return decodeFrame(join([header, await abortable(stream.readExact(length, signal), signal)], FRAME_HEADER_SIZE + length));
}

async function abortable<T>(operation: Promise<T>, signal?: AbortSignal): Promise<T> {
  if (signal === undefined) return operation;
  signal.throwIfAborted();
  let abort: (() => void) | undefined;
  const interrupted = new Promise<never>((_, reject) => {
    abort = (): void => reject(signal.reason ?? new DOMException("aborted", "AbortError"));
    signal.addEventListener("abort", abort, { once: true });
  });
  try {
    return await Promise.race([operation, interrupted]);
  } finally {
    if (abort !== undefined) signal.removeEventListener("abort", abort);
  }
}

function checkedRequestId(value?: bigint): bigint {
  const requestId = value ?? (BigInt(Date.now()) << 16n) | BigInt(Math.floor(Math.random() * 65536));
  if (requestId <= 0n || requestId > (1n << 64n) - 1n) throw new ClientError("request ID must be a nonzero u64");
  return requestId;
}

function join(values: readonly Uint8Array[], length: number): Uint8Array {
  const output = new Uint8Array(length);
  let offset = 0;
  for (const value of values) {
    output.set(value, offset);
    offset += value.byteLength;
  }
  return output;
}

function equal(left: Uint8Array, right: Uint8Array): boolean {
  return left.byteLength === right.byteLength && left.every((value, index) => value === right[index]);
}

File: crates/hyphae-client/src/v2/local.rs
// SPDX-License-Identifier: AGPL-3.0-only

use std::sync::atomic::{AtomicU64, Ordering};

use hyphae_native_product::{ProductOperation, ProductResponse};
use hyphae_native_protocol::{
    API_KEY_AUTH_TRAILER_BYTES, AsyncFrameIo, FrameKind, Hello, ProtocolCapabilities,
    ProvisionalStream, decode_end, decode_failure, decode_welcome, encode_authenticated_hello,
    encode_cancel, encode_hello, encode_product_request, encode_window_update,
};
use tokio::sync::Mutex;

#[cfg(unix)]
use interprocess::local_socket::GenericFilePath;
#[cfg(windows)]
use interprocess::local_socket::GenericNamespaced;

use super::{ClientError, RequestOptions, ResponseFuture, Transport};

const CLIENT_IDENTITY: &str = "hyphae-rust-sdk-v2";
#[cfg(windows)]
const WINDOWS_PIPE_PREFIX: &str = r"\\.\pipe\";

/// Exact `HYPHLCL1` transport over `AF_UNIX` or Windows named pipes.
pub struct LocalTransport {
    endpoint: String,
    client_identity: String,
    api_key: Option<LocalApiKey>,
    state: Mutex<Option<Connection>>,
    next_request_id: AtomicU64,
}

struct LocalApiKey {
    bytes: Vec<u8>,
}

impl LocalApiKey {
    fn new(value: impl AsRef<str>) -> Result<Self, ClientError> {
        let bytes = value.as_ref().as_bytes();
        if bytes.len() != API_KEY_AUTH_TRAILER_BYTES {
            return Err(ClientError::Local(
                "local API-key credential is invalid".to_owned(),
            ));
        }
        Ok(Self {
            bytes: bytes.to_vec(),
        })
    }

    fn expose(&self) -> Result<&str, ClientError> {
        std::str::from_utf8(&self.bytes)
            .map_err(|_| ClientError::Local("local API-key credential is invalid".to_owned()))
    }
}

impl Drop for LocalApiKey {
    fn drop(&mut self) {
        self.bytes.fill(0);
    }
}

impl std::fmt::Debug for LocalTransport {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter
            .debug_struct("LocalTransport")
            .field("endpoint", &self.endpoint)
            .finish_non_exhaustive()
    }
}

struct Connection {
    stream: interprocess::local_socket::tokio::Stream,
    codec: AsyncFrameIo,
    maximum_response_bytes: usize,
    next_stream_id: u32,
}

#[allow(clippy::missing_errors_doc)]
impl LocalTransport {
    /// Configures one platform-local endpoint path or named-pipe identity.
    pub fn new(endpoint: impl Into<String>) -> Result<Self, ClientError> {
        let endpoint = normalize_endpoint(endpoint.into())?;
        if endpoint.is_empty() {
            return Err(ClientError::Local("local endpoint is empty".to_owned()));
        }
        Ok(Self {
            endpoint,
            client_identity: CLIENT_IDENTITY.to_owned(),
            api_key: None,
            state: Mutex::new(None),
            next_request_id: AtomicU64::new(1),
        })
    }

    /// Overrides the bounded identity sent in the native-local handshake.
    pub fn client_identity(mut self, identity: impl Into<String>) -> Result<Self, ClientError> {
        let identity = identity.into();
        if identity.is_empty() || identity.len() > 4 * 1024 {
            return Err(ClientError::Local(
                "local client identity is invalid".to_owned(),
            ));
        }
        self.client_identity = identity;
        Ok(self)
    }

    /// Requires durable Native API-key authentication during local handshake.
    pub fn api_key(mut self, api_key: impl AsRef<str>) -> Result<Self, ClientError> {
        self.api_key = Some(LocalApiKey::new(api_key)?);
        Ok(self)
    }

    async fn connect(&self, handshake_id: u64) -> Result<Connection, ClientError> {
        use interprocess::local_socket::tokio::prelude::*;

        #[cfg(unix)]
        let name = self
            .endpoint
            .as_str()
            .to_fs_name::<GenericFilePath>()
            .map_err(|error| ClientError::Local(error.to_string()))?;
        #[cfg(windows)]
        let name = self
            .endpoint
            .as_str()
            .to_ns_name::<GenericNamespaced>()
            .map_err(|error| ClientError::Local(error.to_string()))?;
        let stream = interprocess::local_socket::tokio::Stream::connect(name)
            .await
            .map_err(|error| ClientError::Local(error.to_string()))?;
        let mut codec = AsyncFrameIo::new(hyphae_native_protocol::DEFAULT_MAX_FRAME_PAYLOAD)
            .map_err(|error| ClientError::Protocol(error.to_string()))?;
        let mut hello = Hello {
            client_identity: self.client_identity.clone(),
            ..Hello::default()
        };
        let mut hello_payload = if let Some(api_key) = &self.api_key {
            hello.capabilities = ProtocolCapabilities::G6_AUTHENTICATED;
            hello.required_capabilities = ProtocolCapabilities::G6_AUTHENTICATED;
            encode_authenticated_hello(&hello, api_key.expose()?)
        } else {
            encode_hello(&hello)
        }
        .map_err(|error| ClientError::Protocol(error.to_string()))?;
        let sent = codec
            .send(
                &mut &stream,
                FrameKind::Hello,
                0,
                handshake_id,
                &hello_payload,
            )
            .await;
        hello_payload.fill(0);
        sent.map_err(|error| ClientError::Local(error.to_string()))?;
        let welcome = codec
            .receive(&mut &stream)
            .await
            .map_err(|error| ClientError::Local(error.to_string()))?
            .ok_or_else(|| ClientError::Local("server closed during handshake".to_owned()))?;
        if welcome.kind == FrameKind::Failure {
            return Err(Box::new(
                decode_failure(&welcome.payload)
                    .map_err(|error| ClientError::Protocol(error.to_string()))?,
            )
            .into());
        }
        if welcome.kind != FrameKind::Welcome
            || welcome.stream_id != 0
            || welcome.request_id != handshake_id
        {
            return Err(ClientError::Protocol(
                "server returned a mismatched welcome frame".to_owned(),
            ));
        }
        let welcome = decode_welcome(&welcome.payload)
            .map_err(|error| ClientError::Protocol(error.to_string()))?;
        if self.api_key.is_some()
            && !welcome
                .capabilities
                .contains(ProtocolCapabilities::API_KEY_AUTH)
        {
            return Err(ClientError::Protocol(
                "server downgraded local API-key authentication".to_owned(),
            ));
        }
        codec = AsyncFrameIo::new(
            usize::try_from(welcome.maximum_frame_payload)
                .map_err(|_| ClientError::Protocol("invalid negotiated frame limit".to_owned()))?,
        )
        .map_err(|error| ClientError::Protocol(error.to_string()))?;
        let maximum_response_bytes = usize::try_from(welcome.initial_window)
            .map_err(|_| ClientError::Protocol("invalid negotiated window".to_owned()))?;
        Ok(Connection {
            stream,
            codec,
            maximum_response_bytes,
            next_stream_id: 1,
        })
    }

    #[allow(clippy::too_many_lines)]
    async fn execute_inner(
        &self,
        operation: ProductOperation,
        options: RequestOptions,
    ) -> Result<ProductResponse, ClientError> {
        let request_id = options
            .request_id
            .unwrap_or_else(|| self.next_request_id.fetch_add(1, Ordering::Relaxed));
        if request_id == 0 {
            return Err(ClientError::Protocol(
                "request identity must be nonzero".to_owned(),
            ));
        }
        if options.cancellation.is_cancelled() {
            return Err(product_error(
                hyphae_native_product::ProductErrorCode::Cancelled,
                request_id,
            ));
        }
        let encoded = encode_product_request(&hyphae_native_protocol::WireRequest {
            operation,
            logical_time_micros: options.logical_time_micros,
            deadline_micros: options.deadline_micros,
            idempotency_token: options.idempotency_token,
            limits: options.limits,
            durability: options.durability,
        })
        .map_err(|error| ClientError::Protocol(error.to_string()))?;
        let mut state = self.state.lock().await;
        if state.is_none() {
            let handshake_id = self.next_request_id.fetch_add(1, Ordering::Relaxed);
            *state = Some(self.connect(handshake_id.max(1)).await?);
        }
        let connection = state.as_mut().ok_or_else(|| {
            ClientError::Local("local connection state is unavailable".to_owned())
        })?;
        let stream_id = connection.next_stream_id;
        connection.next_stream_id = connection
            .next_stream_id
            .checked_add(1)
            .filter(|value| *value != 0)
            .unwrap_or(1);
        let kind = operation_frame_kind(&encoded)?;
        connection
            .codec
            .send(
                &mut &connection.stream,
                kind,
                stream_id,
                request_id,
                &encoded,
            )
            .await
            .map_err(|error| ClientError::Local(error.to_string()))?;

        let maximum = options
            .limits
            .max_response_bytes
            .min(hyphae_native_protocol::MAX_PRODUCT_WIRE_BYTES);
        let mut provisional = ProvisionalStream::new();
        let mut credited = 0_u64;
        loop {
            if options.cancellation.is_cancelled() {
                connection
                    .codec
                    .send(
                        &mut &connection.stream,
                        FrameKind::Cancel,
                        stream_id,
                        request_id,
                        &encode_cancel(1),
                    )
                    .await
                    .map_err(|error| ClientError::Local(error.to_string()))?;
                return Err(product_error(
                    hyphae_native_product::ProductErrorCode::Cancelled,
                    request_id,
                ));
            }
            let mut reader = &connection.stream;
            let received = tokio::select! {
                result = connection.codec.receive(&mut reader) => {
                    result.map_err(|error| ClientError::Local(error.to_string()))?
                }
                () = options.cancellation.cancelled() => {
                    connection.codec.send(
                        &mut &connection.stream,
                        FrameKind::Cancel,
                        stream_id,
                        request_id,
                        &encode_cancel(1),
                    ).await.map_err(|error| ClientError::Local(error.to_string()))?;
                    *state = None;
                    return Err(product_error(
                        hyphae_native_product::ProductErrorCode::Cancelled,
                        request_id,
                    ));
                }
                () = wait_deadline(options.deadline_micros) => {
                    connection.codec.send(
                        &mut &connection.stream,
                        FrameKind::Cancel,
                        stream_id,
                        request_id,
                        &encode_cancel(2),
                    ).await.map_err(|error| ClientError::Local(error.to_string()))?;
                    *state = None;
                    return Err(product_error(
                        hyphae_native_product::ProductErrorCode::DeadlineExceeded,
                        request_id,
                    ));
                }
            };
            let Some(frame) = received else {
                *state = None;
                return Err(ClientError::Local(
                    "server closed before stream completion".to_owned(),
                ));
            };
            if frame.request_id != request_id || frame.stream_id != stream_id {
                *state = None;
                return Err(ClientError::Protocol(
                    "serial client received a mismatched response".to_owned(),
                ));
            }
            match frame.kind {
                FrameKind::Data => {
                    provisional
                        .push(&frame.payload, maximum)
                        .map_err(|error| ClientError::Protocol(error.to_string()))?;
                    credited = credited.saturating_add(
                        u64::try_from(frame.payload.len())
                            .map_err(|_| ClientError::Protocol("response too large".to_owned()))?,
                    );
                    if credited >= u64::try_from(connection.maximum_response_bytes / 2).unwrap_or(1)
                    {
                        connection
                            .codec
                            .send(
                                &mut &connection.stream,
                                FrameKind::WindowUpdate,
                                stream_id,
                                request_id,
                                &encode_window_update(credited)
                                    .map_err(|error| ClientError::Protocol(error.to_string()))?,
                            )
                            .await
                            .map_err(|error| ClientError::Local(error.to_string()))?;
                        credited = 0;
                    }
                }
                FrameKind::End => {
                    let bytes = provisional
                        .complete(
                            decode_end(&frame.payload)
                                .map_err(|error| ClientError::Protocol(error.to_string()))?,
                        )
                        .map_err(|error| ClientError::Protocol(error.to_string()))?;
                    return hyphae_native_protocol::decode_product_response(&bytes)
                        .map_err(|error| ClientError::Protocol(error.to_string()));
                }
                FrameKind::Failure => {
                    return Err(Box::new(
                        decode_failure(&frame.payload)
                            .map_err(|error| ClientError::Protocol(error.to_string()))?,
                    )
                    .into());
                }
                _ => {
                    *state = None;
                    return Err(ClientError::Protocol(
                        "server returned an invalid response frame".to_owned(),
                    ));
                }
            }
        }
    }
}

fn product_error(code: hyphae_native_product::ProductErrorCode, request_id: u64) -> ClientError {
    Box::new(
        hyphae_native_product::ProductError::from_code(code)
            .with_request_id(u128::from(request_id)),
    )
    .into()
}

#[cfg_attr(unix, allow(clippy::unnecessary_wraps))]
fn normalize_endpoint(endpoint: String) -> Result<String, ClientError> {
    #[cfg(unix)]
    {
        Ok(endpoint)
    }
    #[cfg(windows)]
    {
        let bare = endpoint
            .get(..WINDOWS_PIPE_PREFIX.len())
            .filter(|prefix| prefix.eq_ignore_ascii_case(WINDOWS_PIPE_PREFIX))
            .map_or(endpoint.as_str(), |_| {
                &endpoint[WINDOWS_PIPE_PREFIX.len()..]
            });
        if bare.is_empty() || bare.starts_with(r"\\") {
            return Err(ClientError::Local(
                "Windows local endpoint must be a local named-pipe namespace".to_owned(),
            ));
        }
        Ok(bare.to_owned())
    }
}

impl Transport for LocalTransport {
    fn execute(&self, operation: ProductOperation, options: RequestOptions) -> ResponseFuture<'_> {
        Box::pin(self.execute_inner(operation, options))
    }
}

fn operation_frame_kind(encoded: &[u8]) -> Result<FrameKind, ClientError> {
    if encoded.len() < 14 {
        return Err(ClientError::Protocol(
            "encoded request is truncated".to_owned(),
        ));
    }
    Ok(match u16::from_le_bytes([encoded[12], encoded[13]]) {
        2 => FrameKind::Prepare,
        12 => FrameKind::Deallocate,
        _ => FrameKind::Execute,
    })
}

async fn wait_deadline(deadline_micros: Option<i64>) {
    let Some(deadline_micros) = deadline_micros else {
        std::future::pending::<()>().await;
        return;
    };
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or(std::time::Duration::ZERO)
        .as_micros();
    let deadline = u128::try_from(deadline_micros).unwrap_or(0);
    if deadline > now {
        let remaining = deadline - now;
        tokio::time::sleep(std::time::Duration::from_micros(
            u64::try_from(remaining).unwrap_or(u64::MAX),
        ))
        .await;
    }
}

#[cfg(all(test, windows))]
mod windows_tests {
    use super::*;

    #[test]
    fn named_pipe_endpoints_normalize_to_one_bare_namespace() -> Result<(), ClientError> {
        let bare = LocalTransport::new("hyphae-test")?;
        let full = LocalTransport::new(r"\\.\pipe\hyphae-test")?;
        assert_eq!(bare.endpoint, "hyphae-test");
        assert_eq!(full.endpoint, "hyphae-test");
        assert!(LocalTransport::new(r"\\server\pipe\hyphae-test").is_err());
        Ok(())
    }
}
